proposed/0000-UDP-hole-punching.md

```diff
@@ -0,0 +1,191 @@
+to allow for P2P connections of nodes behind NATs in a fully decentralised
+network.
+
+# Motivation
+
+Most users of the Crust library are expected to run the software on home PCs connected
+to the Internet through a router. Routers usually implement some kind of a firewall to
+prevent anyone from the outside Internet being able to connect to the user's PC directly.
+As Crust is a library that creates P2P connections, it should try available techniques
+to circumvent these firewall restrictions and UDP hole punching is one such
+technique.
+
+Additionally, the traditional UDP hole punching process involves a well known
+server to tell each peer their external UDP endpoints. Since our intent
+is a fully decentralised system, this text also describes how nodes to which
+we're already connected can be used as a replacement for such a server.
+
+In the following text the reader is expected to know basic NAT types and how they work.
+The section 'Methods of translation' on [wikipedia](https://en.wikipedia.org/wiki/Network_address_translation)
+gives a nice overview.
+
+# Detailed design
+
+The UDP hole punching process described here consists of two steps:
+
+1. Finding an external mapping of a UDP socket
+2. Do the actual hole punching.
+
+## Finding an external mapping of a UDP socket
+
+Suppose, we have two nodes `A` and `B`, where each one is possibly behind a NAT and
+they want to connect to each other.  Routing needs to send `A` an endpoint of `B` and
+vice versa. But `A` and `B` don’t know their public
+endpoints a priori (they may know their IP addresses from previous connection establishments,
+but not their ports). Each node therefore needs to create a UDP socket and use it to ask some
+other node it is connected to what the public endpoint of the UDP socket is. We’ll refer
+to such helping nodes `C(A)` and `C(B)` (it need not be the same node for `A` and `B`).
+We'll call `U(C(X))` the UDP port on which X can contact `C(X)` in order to find its public
+endpoint. `X` finds out about `U(C(X))` through the initial handhsake, that is, when
+`X` connects to `C(X)`.
+
+In the following text we’ll describe steps taken at node `A`, but the same steps need to
+be taken at node `B`.
+
+For `A` to find out its external address `A` first needs to create a UDP socket `S(A)`
+with an arbitrary local port.  `A` will then use `S(A)` to periodically send
+datagrams to `U(C(A))` containing a MAGIC number and the ID of the request. Each time
+`C(A)` receives such datagram, it sends back a datagram containing public endpoint
+of `S(A)` together with the request ID.
+
+The above protocol shall be initiated from upper layers by calling
+
+    Service::get_mapped_udp_socket(callback: FnBox<(io::Result<MappedUdpSocket>)>)
+
+And the result of such a call shall be an invocation of the callback.
+
+where `MappedUdpSocket` is defined as
+
+    struct MappedUdpSocket {
+        udp_socket: UdpSocket,
+        public_address: SocketAddr, // of node A
+    }
+
```

Comment:
I should change the line "The above protocol shall be initiated from upper layers by calling", as really the plan is to use this function only internally. As it is a function taking a callback, it doesn't really need any token. However, the public API shall indeed use tokens (I just thought I would wait for specific requirements). Though my intention with this RFC is to create a *minimal* requirements for input and output, I believe that will make it easier to *reuse* this function in different contexts. So, say if we went with the idea that the PublicKey is the token, we would move the PublicKey to the callback and use it when creating the event for the user.

However I don't think using PublicKey as a token is a good idea, for these reasons:

1. It is a dependency injection: It forces the user to be aware `crypto::PublicKey`, this suits our needs, but I don't think it applies for everyone.
2. You would not be able to call the function more than once asynchronously for one remote peer.
3. It would be nice to have one token *type* for all async actions, and the type `crypto::PublicKey` would be (at best) strange for actions such as "accept connection" or "map UPnP endpoints".

Though if we have the requirement that we'd want to obfuscate this part of the protocol, then adding `crypto::PublicKey` to the function arguments would be useful, just not as a result token IMO.